// @sa11y/format/dist/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendWcag = exports.A11yResults = exports.A11yResult = exports.exceptionListFilterSelectorKeywords = exports.exceptionListFilter = exports.A11yError = void 0;
var format_1 = require("./format");
Object.defineProperty(exports, "A11yError", { enumerable: true, get: function () { return format_1.A11yError; } });
var filter_1 = require("./filter");
Object.defineProperty(exports, "exceptionListFilter", { enumerable: true, get: function () { return filter_1.exceptionListFilter; } });
Object.defineProperty(exports, "exceptionListFilterSelectorKeywords", { enumerable: true, get: function () { return filter_1.exceptionListFilterSelectorKeywords; } });
var result_1 = require("./result");
Object.defineProperty(exports, "A11yResult", { enumerable: true, get: function () { return result_1.A11yResult; } });
Object.defineProperty(exports, "A11yResults", { enumerable: true, get: function () { return result_1.A11yResults; } });
Object.defineProperty(exports, "appendWcag", { enumerable: true, get: function () { return result_1.appendWcag; } });
