// @sa11y/preset-rules/dist/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkData = exports.rulesData = exports.changesData = exports.WcagMetadata = exports.excludedRules = exports.full = exports.base = exports.extended = exports.getDefaultRuleset = exports.defaultRuleset = exports.getA11yConfig = exports.wcagLevels = exports.priorities = exports.defaultPriority = void 0;
var rules_1 = require("./rules");
Object.defineProperty(exports, "defaultPriority", { enumerable: true, get: function () { return rules_1.defaultPriority; } });
Object.defineProperty(exports, "priorities", { enumerable: true, get: function () { return rules_1.priorities; } });
Object.defineProperty(exports, "wcagLevels", { enumerable: true, get: function () { return rules_1.wcagLevels; } });
Object.defineProperty(exports, "getA11yConfig", { enumerable: true, get: function () { return rules_1.getA11yConfig; } });
var config_1 = require("./config");
Object.defineProperty(exports, "defaultRuleset", { enumerable: true, get: function () { return config_1.defaultRuleset; } });
Object.defineProperty(exports, "getDefaultRuleset", { enumerable: true, get: function () { return config_1.getDefaultRuleset; } });
var extended_1 = require("./extended");
Object.defineProperty(exports, "extended", { enumerable: true, get: function () { return extended_1.extended; } });
var base_1 = require("./base");
Object.defineProperty(exports, "base", { enumerable: true, get: function () { return base_1.base; } });
var full_1 = require("./full");
Object.defineProperty(exports, "full", { enumerable: true, get: function () { return full_1.full; } });
Object.defineProperty(exports, "excludedRules", { enumerable: true, get: function () { return full_1.excludedRules; } });
var wcag_1 = require("./wcag");
Object.defineProperty(exports, "WcagMetadata", { enumerable: true, get: function () { return wcag_1.WcagMetadata; } });
const changes_1 = __importDefault(require("./custom-rules/changes"));
exports.changesData = changes_1.default;
const rules_2 = __importDefault(require("./custom-rules/rules"));
exports.rulesData = rules_2.default;
const checks_1 = __importDefault(require("./custom-rules/checks"));
exports.checkData = checks_1.default;
