// @sa11y/preset-rules/dist/rules.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getA11yConfig = exports.filterRulesByPriority = exports.getPriorityFilter = exports.defaultWcagVersion = exports.defaultPriority = exports.priorities = exports.wcagLevels = void 0;
const common_1 = require("@sa11y/common");
exports.wcagLevels = ['A', 'AA', 'AAA', ''];
exports.priorities = ['P1', 'P2', 'P3', ''];
exports.defaultPriority = 'P3';
exports.defaultWcagVersion = '2.1';
/**
 * Get Priority filter for the default ruleset.
 * When set, only the rules matching the given priority from the default ruleset will be
 * used for the sa11y API and automatic checks.
 */
function getPriorityFilter() {
    const priorityEnv = process.env.SA11Y_RULESET_PRIORITY;
    const priority = priorityEnv && exports.priorities.includes(priorityEnv) ? priorityEnv : '';
    if (priority)
        (0, common_1.log)(`Setting Sa11y rules priority to ${priority}`);
    return priority;
}
exports.getPriorityFilter = getPriorityFilter;
/**
 * Filter given rules and return those matching given priority or .
 */
function filterRulesByPriority(rules, priority = '') {
    // TODO (refactor): Could be simplified by filtering and returning a RuleInfo map
    //  and using the rule keys in the calling function
    const ruleIDs = [];
    const priorityOverride = priority || getPriorityFilter();
    if (!priorityOverride) {
        // if no override is set, return all rules
        ruleIDs.push(...rules.keys());
    }
    else {
        for (const [ruleID, ruleInfo] of rules.entries()) {
            if (priorityOverride === ruleInfo.priority)
                ruleIDs.push(ruleID);
        }
    }
    return ruleIDs.sort();
}
exports.filterRulesByPriority = filterRulesByPriority;
/**
 * Returns config to be used in axe.run() with given rules
 * Ref: https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter
 * @param rules - List of rules to be used in the config
 * @returns A11yConfig with formatted rules
 */
function getA11yConfig(rules) {
    // Note: Making the returned config immutable using Object.freeze() results in
    //  "TypeError: Cannot add property reporter, object is not extensible"
    //  even when no local modifications are made. axe.run() itself seems to be modifying the config object.
    return {
        runOnly: {
            type: 'rule',
            values: filterRulesByPriority(rules),
        },
        // Disable preloading assets as it causes timeouts for audio/video elements
        //  with jest and delays webdriver tests by 2-3x when assets are not found (404)
        //  Ref: https://github.com/dequelabs/axe-core/issues/2528
        preload: false,
        // Types not listed will still show a maximum of one node
        resultTypes: ['violations'],
        reporter: 'no-passes', // return only violation results
    };
}
exports.getA11yConfig = getA11yConfig;
