// @sa11y/common/dist/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCustomRules = exports.processFiles = exports.useCustomRules = exports.useFilesToBeExempted = exports.log = exports.errMsgHeader = exports.getViolations = exports.getAxeRules = exports.axeVersion = exports.axeRuntimeExceptionMsgPrefix = void 0;
var axe_1 = require("./axe");
Object.defineProperty(exports, "axeRuntimeExceptionMsgPrefix", { enumerable: true, get: function () { return axe_1.axeRuntimeExceptionMsgPrefix; } });
Object.defineProperty(exports, "axeVersion", { enumerable: true, get: function () { return axe_1.axeVersion; } });
Object.defineProperty(exports, "getAxeRules", { enumerable: true, get: function () { return axe_1.getAxeRules; } });
Object.defineProperty(exports, "getViolations", { enumerable: true, get: function () { return axe_1.getViolations; } });
var format_1 = require("./format");
Object.defineProperty(exports, "errMsgHeader", { enumerable: true, get: function () { return format_1.errMsgHeader; } });
var helpers_1 = require("./helpers");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return helpers_1.log; } });
Object.defineProperty(exports, "useFilesToBeExempted", { enumerable: true, get: function () { return helpers_1.useFilesToBeExempted; } });
Object.defineProperty(exports, "useCustomRules", { enumerable: true, get: function () { return helpers_1.useCustomRules; } });
Object.defineProperty(exports, "processFiles", { enumerable: true, get: function () { return helpers_1.processFiles; } });
Object.defineProperty(exports, "registerCustomRules", { enumerable: true, get: function () { return helpers_1.registerCustomRules; } });
